appender.go

```diff
@@ -132,150 +179,330 @@ func (a *Appender) AppendRow(args ...driver.Value) error {
-	a.currentRow = 0
-
+	a.currentChunkSize = 0
 	// duckdb_create_data_chunk takes an array of duckdb_logical_type and a column count
-	dataChunk := C.duckdb_create_data_chunk(a.chunkTypes, C.uint64_t(colCount))
-	C.duckdb_data_chunk_set_size(dataChunk, C.uint64_t(a.currentChunkSize))
-
-	// reset the chunkVectors array if they've been previously set
-	if a.chunkVectors != nil {
-		a.chunkVectors = nil
-	}
+	columnTypesPtr := (*C.duckdb_logical_type)(a.columnTypesPtr)
+	dataChunk := C.duckdb_create_data_chunk(columnTypesPtr, C.idx_t(colCount))
+	C.duckdb_data_chunk_set_size(dataChunk, C.duckdb_vector_size())
 
 	for i := 0; i < colCount; i++ {
-		vector := C.duckdb_data_chunk_get_vector(dataChunk, C.uint64_t(i))
+		vector := C.duckdb_data_chunk_get_vector(dataChunk, C.idx_t(i))
 		if vector == nil {
-			return fmt.Errorf("error while appending column %d", i)
+			panic(fmt.Sprintf("error while appending column %d", i))
 		}
-		a.chunkVectors = append(a.chunkVectors, vector)
+		c := &a.columnInfos[i]
+		c.vector = vector
+		c.getChildVectors(vector)
 	}
 
 	a.chunks = append(a.chunks, dataChunk)
 	return nil
 }
 
+func setNull(columnInfo *ColumnInfo, rowIdx C.idx_t) {
+	C.duckdb_vector_ensure_validity_writable(columnInfo.vector)
+	mask := C.duckdb_vector_get_validity(columnInfo.vector)
+	C.duckdb_validity_set_row_invalid(mask, rowIdx)
+}
+
+func setPrimitive[T any](a *Appender, columnInfo *ColumnInfo, rowIdx C.idx_t, value T) {
+	ptr := C.duckdb_vector_get_data(columnInfo.vector)
+	xs := (*[1 << 31]T)(ptr)
+	xs[rowIdx] = value
+}
+
+func setVarchar(a *Appender, columnInfo *ColumnInfo, rowIdx C.idx_t, value string) {
+	str := C.CString(value)
+	C.duckdb_vector_assign_string_element(columnInfo.vector, rowIdx, str)
+	C.free(unsafe.Pointer(str))
+}
+
+func setTime(a *Appender, columnInfo *ColumnInfo, rowIdx C.idx_t, value driver.Value) {
+	var dt C.duckdb_timestamp
+	dt.micros = C.int64_t(value.(time.Time).UTC().UnixMicro())
+	setPrimitive[C.duckdb_timestamp](a, columnInfo, rowIdx, dt)
+}
+
+func setList(a *Appender, columnInfo *ColumnInfo, rowIdx C.idx_t, value driver.Value) {
+	refVal := reflect.ValueOf(value)
+
+	childColumnInfo := columnInfo.columnInfos[0]
+
+	childVectorSize := C.duckdb_list_vector_get_size(columnInfo.vector)
+
+	if refVal.IsNil() {
+		setNull(columnInfo, a.currentChunkSize)
+	}
+
+	// Convert the refVal to []interface{} to be able to iterate over it
+	interfaceSlice := make([]interface{}, refVal.Len())
+	for i := 0; i < refVal.Len(); i++ {
+		interfaceSlice[i] = refVal.Index(i).Interface()
+	}
+
+	// Set the offset of the list vector using the current size of the child vector.
+	// The duckdb list vector contains for each row two values, the offset and the length,
+	// which are located at rowIdx*2 and rowIdx*2+1 respectively.
+	setPrimitive[uint64](a, columnInfo, rowIdx*2, uint64(childVectorSize))
+
+	// Set the length of the list vector
+	setPrimitive[uint64](a, columnInfo, rowIdx*2+1, uint64(refVal.Len()))
```

Comment:
I was wondering - can we use the following struct in `abi.go` for this? And maybe move the comments there?
```Go
type duckdb_list_entry_t struct {
	offset C.idx_t
	length C.idx_t
}
```

Reply:
Alternatively.
C API
```C
typedef struct {
	uint64_t offset;
	uint64_t length;
} duckdb_list_entry;
```
And something like this
```Go
		case Interval:
			val := C.duckdb_interval{
				months: C.int32_t(v.Months),
				days:   C.int32_t(v.Days),
				micros: C.int64_t(v.Micros),
			}
			if rv := C.duckdb_bind_interval(*s.stmt, C.idx_t(i+1), val); rv == C.DuckDBError {
				return errCouldNotBind
			}
```